### ISSUE TITLE
An incorrect match is defined, and the code still compiles successfully

### ISSUE THREAD

**Issue by tswts** [rust-lang/rust#89866]
I defined an enum:
```rust
#[derive(Debug)]
enum Type {
     // =
     Equal,
     //!
     Exclamation,
     // +
     Plus,
     //-
     Minus,
     // *
     Mul,
     // /
     Div,
     // newline
     NewLine,
     // Finish
     Eof,
     // Such as: name or data or is
     Name(String),
     // Such as: "Hello world!"
     String(String),
}

```
Then an incorrect match was defined in the back, and the code passed the compilation:
```rust
             match t {
                 Eof => {break},
             }
```
When I modified the code to this, the error was reported correctly:
```rust
             match t {
                 crate::Type::Eof => {break},
             }
```
cargo version: 1.57.0-nightly (d56b42c54 2021-09-27)
Operating system: windows10
Complete project:
[iss.zip](https://github.com/rust-lang/rust/files/7341874/iss.zip)

In addition, I am not very familiar with the open source community and rust. I am sorry if I did something wrong.
